File: 125-valid-palindrome/125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        # Read all characters. If isAlphaNumeric, then toLower. O(n)
        s_cleaned = ""
        for char in s:
            if char.isalnum():
                s_cleaned += char.lower()
        print(s_cleaned)
        # Reversing is also a O(n) operation
        if s_cleaned == s_cleaned[::-1]:
            return True
        else:
            return False

File: 125-valid-palindrome/125-valid-palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        // 2 pointer problem. O(n)
        int left = 0, right = s.length()-1;
        while (left < right) {
            char head = s[left], tail = s[right];
            if (!isalnum(head)) {
                left++;
                continue;
            }
            if (!isalnum(tail)) {
                right--;
                continue;
            }
            head = tolower(head); tail = tolower(tail);
            if (head != tail) {
                cout << head << " " << tail << endl;
                return false;
            }
            left++; right--;
        }
        return true;
    }
};